app/models/manageiq/providers/vmware/infra_manager/metrics_capture.rb

```diff
@@ -357,37 +339,27 @@ def perf_capture_intervals(mors, interval_name)
     interval_by_mor
   end
 
-  def perf_capture_counters(interval_by_mor)
-    _log.info("Capturing counters...")
-
-    counters_by_mor = {}
-
-    # Query Vim for all of the available metrics and their associated counter info
-    interval_by_mor.each do |mor, interval|
-      counters_by_mor[mor] = self.class.avail_metrics_for_entity(target.ext_management_system, @perf_vim_hist, mor, interval, @perf_intervals[interval.to_s])
-    end
-
-    _log.info("Capturing counters...Complete")
-    counters_by_mor
-  end
-
-  def perf_build_query_params(interval_by_mor, counters_by_mor, start_time, end_time)
+  def perf_build_query_params(interval_by_mor, counter_info, start_time, end_time)
     _log.info("Building query parameters...")
 
     params = []
     interval_by_mor.each do |mor, interval|
-      counters = counters_by_mor[mor]
-      next if counters.empty?
-
       st, et = Metric::Helper.sanitize_start_end_time(interval, @perf_intervals[interval.to_s], start_time, end_time)
 
+      perf_metric_id_set = counter_info.map do |counter_id, _counter_info|
+        {
+          :counterId => counter_id.to_s,
+          :instance  => VIM_PERF_METRIC_ALL_INSTANCES,
+        }
+      end
```

Comment:
Minor, but I think this could be moved outside the loop, as it doesn't rely on anything instance specific anymore.  In fact, can it be moved *WAY* up to where the counter_info is defined, or do we use counter_info for other purposes?

Reply:
We use `counter_info` for `preprocess_data` [here](https://github.com/agrare/manageiq-providers-vmware/blob/6c80834f3b2e1b352698a5b01da3a9446c2edacc/app/models/manageiq/providers/vmware/infra_manager/metrics_capture.rb#L140-L152)

Reply:
We could memoize it similar to what we do with `counter_info_by_id` then just do
```
param = {
  ...
  :metricId => self.class.perf_metric_ids(..)
}
```

but I don't think this is really expensive to compute.

You're right though it should minimally be moved out of that loop to the top of `perf_build_query_params`